### PR TITLE
SPT-1903: Create Dummy FE Screen

### PR DESCRIPTION
**Pull request by mislam246**
### What changed
- Added dummy front end screen
- Added logic to route users from the `/authorize` to the `/confirm-details` endpoint
- Added logic to route users back to the `redirect-uri` passed in from the client, including the auth code and state in the query parameters

### Why did it change
This ticket is required as part of the initial skeleton work for Phase 3, specifically adding a dummy page to SIS front end to accompany the creation of the new SIS endpoints.

### Issue tracking
[SPT-1903](https://govukverify.atlassian.net/browse/SPT-1903)

## Testing
Tested on: https://dev-mi.reuse-identity.dev.account.gov.uk/authorize?response_type=bearer&state=state-id&client_id=client-id&redirect_uri=https://api.example.com

### Checklist for developers

- [x]  The PR description is filled out and the PR title includes the story reference
- [ ]  If there will be further PRs related to the story, this is highlighted in the PR description what is to come
- [x]  There are no merge, WIP or reversion commits included in the PR (reversion commits allowed if reverting previously merged code)
- [x]  All commits include story reference, a distinct title and a body listing changes
- [x]  All commits are signed
- [ ]  All commits contain a `Co-authored-by` line where pairing has taken place
- [x]  All changes in this PR are related to the story
- [x]  The changes have been fully tested in the dev environment
- [x]  There are unit and/or integration tests matching story acceptance criteria (where applicable)
- [ ]  Any feature flags are correctly set for each target environment
- [ ]  Any related configuration changes have been merged and have landed in the target environment(s)
- [ ]  Any related platform changes have been merged and have been applied in the target environment(s)
- [ ]  This change will deploy with zero downtime (consider the blue/green nature of our deployments)
- [ ]  Any Sonarcloud issues are addressed or dismissed with a valid reason
- [ ]  Any exceptions to any of the above statements are documented in the description and agreed with the reviewer

### Checklist for reviewers

- [ ]  The above statements are all true
- [ ]  The change meets the acceptance criteria set out in the story
- [ ]  There are no missing test scenarios
- [ ]  The code is readable and understandable
- [ ]  The code is maintainable and does not present any notable technical debt
- [ ]  There are no outstanding Sonarcloud issues, and you agree with any dismissal reasons

[SPT-1903]: https://govukverify.atlassian.net/browse/SPT-1903?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by umaar — COMMENTED** [review]
I noticed a lot of the feedback I left here: https://github.com/govuk-one-login/ipv-reuse-service-stubs/pull/214 also applies to this PR. Would you be ok with taking on those comments from that PR and applying it here without me having to copy and paste it across?

File: src/handlers/confirm-details/confirm-details-handler.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
+import nunjucks from "nunjucks";
+import confirmDetailsTemplate from "./index.njk";
```
Comment (umaar) [inline src/handlers/confirm-details/confirm-details-handler.ts]: See this comment: https://github.com/govuk-one-login/ipv-reuse-service-stubs/pull/214/changes#r3131166915 which would also apply to this PR (ignore the `express` stuff).

File: src/handlers/confirm-details/confirm-details-submission-handler.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
+import logger from "../../commons/logger";
+
+const extractEventData = async (eventBody: string[]) => {
```
Comment (umaar) [inline src/handlers/confirm-details/confirm-details-submission-handler.ts]: See this comment: https://github.com/govuk-one-login/ipv-reuse-service-stubs/pull/214/changes#r3131168749 tl;dr use `URLSearchParams`. It will help simplify this code, and is built into Node.js, and will be less code for us to maintain/for something to go wrong

File: src/handlers/govuk-resources-uploader/govuk-resources-uploader.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
+import {
```
Comment (umaar) [inline src/handlers/govuk-resources-uploader/govuk-resources-uploader.ts:1]: @chris-dco is coming up with a solution to avoid the [duplication](https://github.com/govuk-one-login/ipv-reuse-service-stubs/blob/main/ais-producer-stub/src/govuk-resources-uploader/govuk-resources-uploader.ts).

_Edit_: Tech lead [decision](https://gds.slack.com/archives/C03EMG4LNGK/p1777025696559599?thread_ts=1776958259.135169&cid=C03EMG4LNGK) is to leave this as-is for now.

File: src/handlers/confirm-details/confirm-details-handler.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      body: nunjucks.renderString(confirmDetailsTemplateContents, {
+        assetPath: "./assets",
+        rootPath: ".",
+        code,
+        redirect_uri,
+        state,
+      }),
+      headers: {
+        "content-type": "text/html",
+        code: code,
```
Comment (umaar) [inline src/handlers/confirm-details/confirm-details-handler.ts]: Interesting what would be the reason for this? Just wanted to check it's correct and intentional? Who/what consumes `code` in the header like this?

Also `code: code` can be shortened to `code`.

File: src/handlers/confirm-details/confirm-details-submission-handler.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  for (const value of eventBody) {
+    const key = value.substring(0, value.indexOf("="));
+    const start = value.indexOf("=") + 1;
+    obj[key] = value.slice(start);
+  }
+  return obj;
+};
+
+export const lambdaHandler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
+  const valuesArray = event.body!.split("&");
```
Comment (umaar) [inline src/handlers/confirm-details/confirm-details-submission-handler.ts]: Do check out [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) which can help here also. FYI if ever in JS you find yourself writing custom regular expressions, or doing lots of `indexOf`'s, there's normally (but not always) a more declarative approach available in vanilla JS. In this case, `URLSearchParams` should be able to alleviate most of this.

File: src/handlers/confirm-details/index.njk [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                                text: message
+                            }
+                        }
+                    ]
+                }) }}
+            </div>
+        {% else %}
+            <div class="govuk-grid-column-two-thirds">
+                <form id="submitForm" action="/confirm-details" method="post">
+                    <input type="hidden" id="redirectUri" name="redirectUri" value="{{redirect_uri}}""></input>
```
Comment (umaar) [inline src/handlers/confirm-details/index.njk]: Looks like there's an extra quote `value="{{redirect_uri}}""` which ironically could exacerbate XSS implications.

File: src/handlers/confirm-details/index.njk [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                    <input type="hidden" id="code" name="code" value="{{code}}""></input>
```
Comment (umaar) [inline src/handlers/confirm-details/index.njk]: [Void elements](https://developer.mozilla.org/en-US/docs/Glossary/Void_element) should avoid end tags. This works though because HTML is forgiving, but we should follow the standard/spec

File: src/handlers/confirm-details/index.njk [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                            value: {
+                                text: message
+                            }
+                        }
+                    ]
+                }) }}
+            </div>
+        {% else %}
+            <div class="govuk-grid-column-two-thirds">
+                <form id="submitForm" action="/confirm-details" method="post">
```
Comment (umaar) [inline src/handlers/confirm-details/index.njk:55]: Two points:

- Have we checked for proper escaping and making sure we're not vulnerable to XSS? e.g. `value="{{redirect_uri}}""`
- The [GDS Way](https://gds-way.digital.cabinet-office.gov.uk/manuals/security-overview-for-websites.html) security section includes recommendations (*must* rather than *should*) for what security headers need adding onto our website. Are we already using them and if not, I feel they should be added in where applicable.

File: src/handlers/confirm-details/index.njk [hunk 9]
```diff
@@ -0,0 +1,76 @@
+{% from "govuk/components/button/macro.njk" import govukButton %}
+{% from "govuk/components/input/macro.njk" import govukInput %}
+{% from "govuk/components/radios/macro.njk" import govukRadios %}
```
Comment (umaar) [inline src/handlers/confirm-details/index.njk]: Maybe a silly question but are all these components being used? Noticed some are being imported but unused? It's better to only import things that are being used

File: src/handlers/confirm-details/index.njk [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+{% from "govuk/components/error-summary/macro.njk" import govukErrorSummary %}
+{% from "govuk/components/panel/macro.njk" import govukPanel %}
+{% from "govuk/components/summary-list/macro.njk" import govukSummaryList %}
+
+{% extends "govuk/template.njk" %}
+{% block head %}
+    <link rel="stylesheet" href="{{ rootPath | default("", true) }}/govuk-frontend.min.css">
+    <base href="./confirm-details" />
```
Comment (umaar) [inline src/handlers/confirm-details/index.njk:10]: How important is this, is it possible to avoid? This can have subtle, hard-to-debug effects on *all resources*. Totally understand if this can't be avoided, but thought to ask as I've had issues with the base tag in the past. Sometimes, absolute references, or references specified from the root level path, can be better. That might not be practical with our setup though?

No need to spend any real time on this one though. `<base>` is totally fair to use in HTML. It's only if there's an immediately obvious alternative technique that I'd suggest to remove the `<base>` tag.

File: template.yaml [hunk 11]
```diff
@@ -997,6 +998,246 @@ Resources:
...
+        - Key: Name
+          Value: !Sub "/aws/lambda/${UploadGovUKFrontEndResourcesLambda}"
+
+  GovUKAssetsBucket:
+    # checkov:skip=CKV_AWS_18: "Ensure the S3 bucket has access logging enabled"
+    # checkov:skip=CKV_AWS_53: "Ensure S3 bucket has block public ACLs enabled"
+    # checkov:skip=CKV_AWS_54: "Ensure S3 bucket has block public policy enabled"
+    # checkov:skip=CKV_AWS_21: "Ensure the S3 bucket has versioning enabled"
+    # checkov:skip=CKV_AWS_55: "Ensure S3 bucket has ignore public ACLs enabled"
+    # checkov:skip=CKV_AWS_56: "Ensure S3 bucket has RestrictPublicBuckets enabled"
```
Comment (umaar) [inline template.yaml]: Just wanted to confirm this is all intentional? Are we potentially disabling valid checks, for example `Ensure S3 bucket has block public ACLs enabled`....if API Gateway is the resource accessing the bucket, then would such a check still be valid?

Disclaimer: I am not an AWS person so maybe someone else can advise also. My naive thought is that the bucket shouldn't be public since traffic is routed through API Gateway, and Gateway can access the bucket with the correct permission.

File: src/handlers/confirm-details/confirm-details-handler.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
```
Comment (umaar) [inline src/handlers/confirm-details/confirm-details-handler.ts:1]: I noticed `src/handlers/confirm-details/confirm-details-submission-handler.ts` _does_ have a test file here `src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts` but then `src/handlers/confirm-details/confirm-details-handler.ts` (this file) _does not_ have a test file. Should we add tests for this file?

File: src/handlers/confirm-details/confirm-details-handler.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+const confirmDetailsTemplateContents = fs.readFileSync(confirmDetailsTemplate, "utf8");
+
+export type ConfirmDetailsQueryStringParameters = {
+  code: string;
+  redirect_uri: string;
+  state: string;
+};
+
+export const lambdaHandler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
+  const { code, redirect_uri, state } = event.queryStringParameters as ConfirmDetailsQueryStringParameters;
```
Comment (umaar) [inline src/handlers/confirm-details/confirm-details-handler.ts:16]: Ignoring TypeScript, think we need a runtime check to check that these exist, otherwise they'd be `undefined` and would silently make their way into the template. Without having thought about this a lot, guess we'd also need to return an appropriate response in the case they don't exist?

File: src/handlers/confirm-details/confirm-details-handler.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        "content-type": "text/html",
+        code: code,
+      },
+    };
+  } catch (err) {
+    logger.error(`Error in lambdaHandler event: ${err}`);
+    return {
+      statusCode: 500,
+      body: JSON.stringify({
+        message: err,
```
Comment (umaar) [inline src/handlers/confirm-details/confirm-details-handler.ts]: Just wondering if you got a chance to check what this actually returns. Because if `err` is an [Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) object, I think this won't show the contents of that error message?

File: samconfig.toml [hunk 15]
```diff
@@ -3,8 +3,8 @@ version = 0.1
 [default.deploy.parameters]
 resolve_s3 = true
 region = "eu-west-2"
-confirm_changeset = true
+confirm_changeset = false
```
Comment (umaar) [inline samconfig.toml:6]: Just confirming this is intentional?

File: src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+  const response = await lambdaHandler(event);
+  expect(response).toStrictEqual({
+    statusCode: 302,
+    body: "",
+    headers: {
+      Location: "https://api.example.com/?code=abc123&state=test-state-id",
+    },
+  });
+});
```
Comment (umaar) [inline src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts:19]: Maybe my eyes deceive me but I think this is identical to the test below

File: src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
+import { APIGatewayEventRequestContextWithAuthorizer, APIGatewayProxyEvent } from "aws-lambda";
```
Comment (umaar) [inline src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts:1]: Currently, the test cases feel quite limited. I might have missed something but I think there are no negative tests. E.g. missing body, missing fields, etc. It might be good to add those in.

File: openAPI/public-api.yaml [hunk 18]
```diff
@@ -32,6 +32,84 @@ paths:
...
+          Fn::GetAtt: ConfirmDetailsLambdaRestApiRole.Arn
+        type: aws_proxy
+    post:
+      parameters:
+        - in: query
+          name: redirect_uri
+          schema:
+            type: string
+      responses:
+        "200":
```
Comment (umaar) [inline openAPI/public-api.yaml]: Is this correct? Good chance I'm getting confused but I think the tests say this should be a 302 possibly

File: openAPI/public-api.yaml [hunk 19]
```diff
@@ -32,6 +32,84 @@ paths:
...
+          default:
+            statusCode: "200"
+            responseParameters:
+              method.response.header.Cache-Control: "'public, max-age=10'"
+              method.response.header.Content-Type: integration.response.header.Content-Type
+              method.response.header.Content-Length: integration.response.header.Content-Length
+              method.response.header.Timestamp: integration.response.header.Date
+        requestParameters:
+          integration.request.path.item: method.request.path.item
+        # passthroughBehavior: when_no_match
```
Comment (umaar) [inline openAPI/public-api.yaml]: Do we need this commented out code?

File: lambda-layers/Makefile [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    $(if $(patsubst /,,$(2)),\
+        $(call find-up,$(1),$(patsubst %/,%,$(dir $(2)))),\
+    )\
+)
+
+PROJECT_ROOT := $(strip $(call find-up,$(TARGET_DIR),$(MAKEFILE_LIST)))
+
+build-GovUKFrontEndLayer:
+	@if [ ! -n "$(PROJECT_ROOT)" ]; then \
+		echo "The node_modules folder was no found."; \
```
Comment (umaar) [inline lambda-layers/Makefile]: Typo: `no found`

File: .github/workflows/build.yml [hunk 21]
```diff
@@ -38,6 +38,10 @@ jobs:
           node-version-file: ".nvmrc"
           cache: npm
 
+      - name: Install Dependencies
+        working-directory: ./
```
Comment (umaar) [inline .github/workflows/build.yml]: Wouldn't this be the default already? Wondering if this is needed?

File: template.yaml [hunk 22]
```diff
@@ -997,6 +998,246 @@ Resources:
...
+      BuildProperties:
+        EntryPoints:
+          - src/handlers/confirm-details/confirm-details-submission-handler.ts
+        Target: es2022
+        Minify: true
+        Sourcemap: true
+
+  ConfirmDetailsLambdaRestApiRole:
+    Type: AWS::IAM::Role
+    Condition: IsNotProduction
```
Comment (umaar) [inline template.yaml:1179]: An AWS person will have better insight here but if we're not wanting to create AWS resources in prod, then do the other resources also need `IsNotProduction`, right now some of the other resources do not have this condition

File: template.yaml [hunk 23]
```diff
@@ -997,6 +998,246 @@ Resources:
...
+          - Effect: Allow
+            Principal:
+              Service: lambda.amazonaws.com
+            Action:
+              - "sts:AssumeRole"
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess
+        - arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole
+
+  ConfirmDetailsLambda:
```
Comment (umaar) [inline template.yaml:1143]: Treat this with a grain of salt, people with more experience in AWS will know more. But I was looking at the Authorization Lambda and noticed we might be missing some properties e.g. `Format: esm` / `Platform: node` and maybe more. Wonder if those are needed?

File: template.yaml [hunk 24]
```diff
@@ -997,6 +998,246 @@ Resources:
...
+      Layers:
+        - !Ref GovUKFrontEndLayer
+    Metadata:
+      BuildMethod: esbuild
+      BuildProperties:
+        UseNpmCi: True
+        EntryPoints:
+          - src/handlers/govuk-resources-uploader/govuk-resources-uploader.ts
+        Sourcemap: true
+        Target: "es2020"
```
Comment (umaar) [inline template.yaml]: Sometimes we use `es2022`, best to be consistent.

**Comment by umaar** [comment]
Nice work, looks really good! I left some comments that maybe @chrisclayson  could skim through also

File: openAPI/public-api.yaml (hunk 18 above)
Comment (chrisclayson) [reply openAPI/public-api.yaml]: Good catch ... this will always be a redirect

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 13 above)
Comment (chrisclayson) [reply src/handlers/confirm-details/confirm-details-handler.ts:16]: Indeed ... we will be adding an error page, so any errors should be redirected to there ... but doesn't exist yet. I think, throw for the time being, and when we add the error page we can add the redirect.

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 4 above)
Comment (chrisclayson) [reply src/handlers/confirm-details/confirm-details-handler.ts]: We don't need a `code` header.

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 14 above)
Comment (chrisclayson) [reply src/handlers/confirm-details/confirm-details-handler.ts]: Good point. Also, we may want to avoid returning the raw content of an error (such as stack trace) in the response, we should definitely log it, but not return it.

Comment (chrisclayson) [reply src/handlers/confirm-details/confirm-details-handler.ts]: Though, as above, we're likely to replace this with a redirect to an error page.

File: template.yaml (hunk 22 above)
Comment (chrisclayson) [reply template.yaml:1179]: Correct ... I think `sam validate` should have picked this up as you have a conditional on the role but not the lambda.

_Really_ the only thing we don't want in production is the API Gateway (i.e. the front door that will let people through), I'm OK with the lambdas being deployed as they went do anything/cost us anything.

File: template.yaml (hunk 23 above)
Comment (mislam246) [reply template.yaml:1143]: @chrisclayson / @chris-dco do we need these additional properties on this Lambda?

File: template.yaml (hunk 11 above)
Comment (mislam246) [reply template.yaml]: I'm not entirely sure. @chrisclayson / @chris-dco which of these do you think we should keep or remove?

File: src/handlers/confirm-details/confirm-details-submission-handler.ts (hunk 2 above)
Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-submission-handler.ts]: That makes it much simpler, thanks for the suggestion. I've updated this to use `URLSearchParams` now.

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 4 above)
Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-handler.ts]: Initially had this for some local testing. Removed the header now.

File: src/handlers/confirm-details/confirm-details-submission-handler.ts (hunk 5 above)
Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-submission-handler.ts]: Updated to use `URLSearchParams` as above.

File: src/handlers/confirm-details/index.njk (hunk 6 above)
Comment (mislam246) [reply src/handlers/confirm-details/index.njk]: Removed the extra quotes for all three inputs.

File: src/handlers/confirm-details/index.njk (hunk 7 above)
Comment (mislam246) [reply src/handlers/confirm-details/index.njk]: Good suggestion. I've removed the end tags for these inputs.

File: src/handlers/confirm-details/index.njk (hunk 8 above)
Comment (mislam246) [reply src/handlers/confirm-details/index.njk:55]: From what I've read, character escaping is enabled by default in Nunjucks. As we discussed, the testing of security headers is possibly a wider piece of work.

File: src/handlers/confirm-details/index.njk (hunk 9 above)
Comment (mislam246) [reply src/handlers/confirm-details/index.njk]: I've replaced the HTML inputs with the `govukInput` component and removed the `govukRadios` import. The rest of the components were being used originally.

File: src/handlers/confirm-details/index.njk (hunk 10 above)
Comment (mislam246) [reply src/handlers/confirm-details/index.njk:10]: I've faced problems getting it to work without the `<base>` tag. Given this will be revisited when we work on the front-end, is it okay to keep as is for now?

File: template.yaml (hunk 11 above)
Comment (mislam246) [reply template.yaml]: As discussed, I've now removed all but the first line and added the `VersioningConfiguration` and `PublicAccessBlockConfiguration` properties to the bucket in the same way as this [example in SPOT](https://github.com/govuk-one-login/ipv-spot-key-management/blob/447b1d89cf9b4e2aef284d7b2002315104332b47/template.yaml#L974).

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 13 above)
Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-handler.ts:16]: Added a check on these parameters and thrown an error if any are undefined.

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 14 above)
Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-handler.ts]: It returns an object containing the error message. I've changed this to not return the error and only log it.

File: samconfig.toml (hunk 15 above)
Comment (mislam246) [reply samconfig.toml:6]: I initially changed this to make dev deployments quicker for testing. If you feel it's safer, I'm happy to undo the change.

File: src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts (hunk 16 above)
Comment (mislam246) [reply src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts:19]: That test is no longer needed, I've removed it.

File: openAPI/public-api.yaml (hunk 18 above)
Comment (mislam246) [reply openAPI/public-api.yaml]: Changed this response code to `302`.

File: openAPI/public-api.yaml (hunk 19 above)
Comment (mislam246) [reply openAPI/public-api.yaml]: Not required, I've removed it.

File: lambda-layers/Makefile (hunk 20 above)
Comment (mislam246) [reply lambda-layers/Makefile]: Fixed the typo.

File: template.yaml (hunk 22 above)
Comment (mislam246) [reply template.yaml:1179]: Added the `IsNotProduction` condition on the other resources.

File: template.yaml (hunk 23 above)
Comment (mislam246) [reply template.yaml:1143]: As discussed, I've added the additional properties (`UseNpmCi`, `Format`, `Platform`, `Target`, `OutExtension`, and `Banner`) with the same values as this [SPOT function](https://github.com/govuk-one-login/ipv-spot/blob/a5663caced075c767606b1e1c1685c2c9dd68f61/template-spot-functions.yaml#L425).

File: template.yaml (hunk 24 above)
Comment (mislam246) [reply template.yaml]: Changed this to use `es2022` instead.

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 12 above)
Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-handler.ts:1]: Added the `src/handlers/confirm-details/confirm-details-handler.test.ts` file with additional tests.

File: src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts (hunk 17 above)
Comment (mislam246) [reply src/handlers/confirm-details/tests/confirm-details-submission-handler.test.ts:1]: I've added some tests to check an error is thrown when there are missing required URL query parameters in the request body.

File: .github/workflows/build.yml (hunk 21 above)
Comment (mislam246) [reply .github/workflows/build.yml]: Tested without this and it's working 👍

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 1 above)
Comment (umaar) [reply src/handlers/confirm-details/confirm-details-handler.ts]: @mislam246 I did some research into this and have fixed this in fa7109055cdff15be1466af90fc38f402a4f6328. We spoke yesterday about the _what_ so I won't repeat that here, but in case it's helpful, I wanted to explain the _why_, so it's clearer:

The problem with creating a custom `.njk=file` loader in esbuild: esbuild is a JavaScript/TypeScript bundler. Nunjucks templates are static text files that have nothing to do with JavaScript compilation. 

Creating a custom esbuild `.njk=file` loader to get a template into the Lambda artifact forces a chain of workarounds:

1. TypeScript module declaration (declare module `"*.njk"`): needed to make `import x from "./index.njk"` type-check, since TypeScript doesn't natively understand .njk imports.
2. Non-standard import pattern: `import template from "./index.njk"` only resolves to a file path because of the bespoke esbuild loader, not because of any standard JS/TS semantics.
3. `fs.readFileSync` + `nunjucks.renderString()`: a roundabout reimplementation of what `nunjucks.render()` already does with its built-in [FileSystemLoader](https://mozilla.github.io/nunjucks/api.html#filesystemloader).
4. We had to write a custom Vitest plugin (`nunjucks-raw-loader`) since tests can't resolve `.njk` imports either, so a custom Vite transform plugin was needed to make tests pass.
5. Content-hashed filenames: I believe esbuild's file loader outputs stuff like `index-abc123.njk` (content-hashed), so even the runtime filename is unpredictable.
6. ESM compatibility shim: when converting to ESM output, a Banner with `import { createRequire } from 'module'; const require = createRequire(import.meta.url);` was needed.

The proper way of doing this is to just copy the template directly. We then get a simpler API `nunjucksEnv.render("index.njk", context);`.

What this removes:
- `src/types.d.ts`: the declare module `"*.njk"` TypeScript declaration
- `.njk=file` esbuild loader config
- `fs.readFileSync` + `nunjucks.renderString()` pattern
- `nunjucks-raw-loader` custom vitest plugin and assetsInclude config
- The ESM Banner shim
- `import confirmDetailsTemplate from "./index.njk"`: no importing templates in JS

I've deployed to dev, and confirmed it's working. Happy to explain more if any questions

Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-handler.ts]: Thanks for the fix and the detailed explanation 👍

File: .github/workflows/build.yml
Comment (chris-dco) [inline .github/workflows/build.yml:56]: Add this to fix your build error: 
```suggestion
            openAPI/
            lambda-layers/
```

Comment (mislam246) [reply .github/workflows/build.yml:56]: Thanks. For future reference, the missing GitHub permissions also needed adding to `gha-aws-oidc-provider-template` and then running the `./apply-role-changes-dev.sh` script.

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 13 above)
Comment (umaar) [reply src/handlers/confirm-details/confirm-details-handler.ts:16]: A check has been added, but it's inconsistent: If state is missing from the query string, it would be `undefined`, which is not `=== ""`.  

Also, relying on [loose equality](https://eslint.org/docs/latest/rules/eqeqeq) to catch both null and undefined is a subtle JS quirk that's easy to misread or break. We should probably apply that eslint rule in the future.

For now, to fix this, you could just do `if (!code || !redirect_uri || !state) {` which also catches empty strings. Maybe worth thinking about whether or not the unit tests would've caught that - no harm in adding a test in for that.

File: samconfig.toml (hunk 15 above)
Comment (umaar) [reply samconfig.toml:6]: Oh should be ok I think, was just to check there was a reason

File: src/handlers/confirm-details/confirm-details-handler.ts (hunk 13 above)
Comment (mislam246) [reply src/handlers/confirm-details/confirm-details-handler.ts:16]: Makes sense, thanks for clarifying. I've refactored it as suggested and added another test to check for empty strings.

**Review by umaar — DISMISSED** [review]

**Review by chrisclayson — APPROVED** [review]
